### ISSUE TITLE
Why do you get only 61.96% AP50 for YOLOv4-512x512 instead of 64.9% AP50 that I can get by using Darknet?

### ISSUE THREAD

**Issue by AlexeyAB** [hunglc007/tensorflow-yolov4-tflite#165]
@hunglc007 Hi,

Why do you get only `61.96%` AP50 for YOLOv4-512x512 https://github.com/hunglc007/tensorflow-yolov4-tflite#map50-on-coco-2017-dataset instead of `64.9%` AP50 that I can get by using Darknet? https://github.com/AlexeyAB/darknet/wiki/How-to-evaluate-accuracy-and-speed-of-YOLOv4

Did you use https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.weights for evaluation or train your own model for this case?

**Comment by AlexeyAB** [comment]
There seems to be inconsistency in detection using the Darknet and Tensorflow. TF detects fewer objects and with less confidence.

It seems to find the reason you need a layer-by-layer comparison of the feature map TF vs Darknet.

| Model | Darknet | hunglc007/tensorflow-yolov4-tflite (pb, tflite, tfile16) |
|---|---|---|
| YOLOv4 | ![yolo_v4](https://user-images.githubusercontent.com/4096485/88479835-ade9ac00-cf5a-11ea-9d4f-32b4cc3dd0f2.jpg) | ![tf_v4_pb](https://user-images.githubusercontent.com/4096485/88479853-c1951280-cf5a-11ea-8af1-7ef7737fdd49.png) |
| YOLOv4-tiny | ![yolo_v4_tiny](https://user-images.githubusercontent.com/4096485/88479860-cb1e7a80-cf5a-11ea-9759-acc69d7f8c96.jpg) | ![tf_v4_tiny_pb](https://user-images.githubusercontent.com/4096485/88479865-d2de1f00-cf5a-11ea-8148-cd852492d566.png) |

----

**Original image:**
![test](https://user-images.githubusercontent.com/4096485/88479880-028d2700-cf5b-11ea-876d-b33a9c7b8b90.png)

**Comment by robisen1** [comment]
> There seems to be inconsistency in detection using the Darknet and Tensorflow. TF detects fewer objects and with less confidence.
> 
> It seems to find the reason you need a layer-by-layer comparison of the feature map TF vs Darknet.
> 
> Model	Darknet	hunglc007/tensorflow-yolov4-tflite (pb, tflite, tfile16)
> YOLOv4	![yolo_v4](https://user-images.githubusercontent.com/4096485/88479835-ade9ac00-cf5a-11ea-9d4f-32b4cc3dd0f2.jpg)	![tf_v4_pb](https://user-images.githubusercontent.com/4096485/88479853-c1951280-cf5a-11ea-8af1-7ef7737fdd49.png)
> YOLOv4-tiny	![yolo_v4_tiny](https://user-images.githubusercontent.com/4096485/88479860-cb1e7a80-cf5a-11ea-9759-acc69d7f8c96.jpg)	![tf_v4_tiny_pb](https://user-images.githubusercontent.com/4096485/88479865-d2de1f00-cf5a-11ea-8148-cd852492d566.png)
> **Original image:**
> ![test](https://user-images.githubusercontent.com/4096485/88479880-028d2700-cf5b-11ea-876d-b33a9c7b8b90.png)

Would you recommend and better TensorFlow implementation. my better i mean better detection and faster processing.

**Comment by hhk7734** [comment]
![result (1)](https://user-images.githubusercontent.com/21096989/90586272-bf4c6e00-e211-11ea-88b8-8753f3ff46f3.png)

I am training yolov4-tiny with activation set to relu. The inference result using weights(mAP50-29%) is as above. I don't know if this will help.

**Comment by hhk7734** [comment]
![yolov4-tiny](https://user-images.githubusercontent.com/21096989/90595403-ca11fd80-e227-11ea-9720-ee3172a0ce5b.png)

Using yolov4-tiny.weights, I trained 300 times with batch_size=32. The result is as above.

**Comment by robisen1** [comment]
> ![result (1)](https://user-images.githubusercontent.com/21096989/90586272-bf4c6e00-e211-11ea-88b8-8753f3ff46f3.png)
> 
> I am training yolov4-tiny with activation set to relu. The inference result using weights(mAP50-29%) is as above. I don't know if this will help.

Have you made any other changes besides relu? I would be interested in looking at the code if your able/willing to share.

**Comment by hhk7734** [comment]
Ref
- https://wiki.loliot.net/docs/etc/project/yolov4/yolov4-training
- https://github.com/hhk7734/tensorflow-yolov4

```python
from yolov4.tf import YOLOv4

yolo = YOLOv4()

yolo.classes = "coco.names"

yolo.make_model(activation1="relu")
yolo.load_weights("yolov4-tiny-relu.weights", weights_type="yolo")

yolo.inference(media_path="kite.jpg")

yolo.inference(media_path="road.mp4", is_image=False)
```

**Comment by robisen1** [comment]
Thank you.

**Comment by vqbang** [comment]
@AlexeyAB @hunglc007 Do you have anything new about the conversion between darknet and tensorflow? I have inconsistent results when converting from yolov3 darknet to tensorflow with this repo too. Currently, there aren't any repos that can convert the exact result from darknet to tensorflow as my research.

**Comment by Fetulhak** [comment]
> @AlexeyAB @hunglc007 Do you have anything new about the conversion between darknet and tensorflow? I have inconsistent results when converting from yolov3 darknet to tensorflow with this repo too. Currently, there aren't any repos that can convert the exact result from darknet to tensorflow as my research.

@vqbang did you get any solution?

**Comment by zz100chan** [comment]
> > @AlexeyAB @hunglc007 Do you have anything new about the conversion between darknet and tensorflow? I have inconsistent results when converting from yolov3 darknet to tensorflow with this repo too. Currently, there aren't any repos that can convert the exact result from darknet to tensorflow as my research.
> 
> @vqbang did you get any solution?

Yes I would also like to hear what you get too!

**Comment by Fetulhak** [comment]
@zz100chan unfortunately I did not get the solution. Maybe there are other heuristics in the Darknet version which may be the reason for its good detection performance. The tenserflow version of YOLO detectors miss a lot of detections. I don't know why? Even I tried the OpenCV Yolo DNN but still not as good as Darknet.
